packages/events/ReactGenericBatching.js

```diff
@@ -14,35 +17,53 @@ import {restoreStateIfNeeded} from './ReactControlledComponent';
 // scheduled work and instead do synchronous work.
 
 // Defaults
-let fiberBatchedUpdates = function(fn, bookkeeping) {
+let _batchedUpdates = function(fn, bookkeeping) {
   return fn(bookkeeping);
 };
+let _interactiveUpdates = function(fn, a, b) {
+  return fn(a, b);
+};
+let _flushInteractiveUpdates = function() {};
 
-let isNestingBatched = false;
+let isBatching = false;
 export function batchedUpdates(fn, bookkeeping) {
-  if (isNestingBatched) {
+  if (isBatching) {
     // If we are currently inside another batch, we need to wait until it
-    // fully completes before restoring state. Therefore, we add the target to
-    // a queue of work.
-    return fiberBatchedUpdates(fn, bookkeeping);
+    // fully completes before restoring state.
+    return fn(bookkeeping);
   }
-  isNestingBatched = true;
+  isBatching = true;
   try {
-    return fiberBatchedUpdates(fn, bookkeeping);
+    return _batchedUpdates(fn, bookkeeping);
   } finally {
     // Here we wait until all updates have propagated, which is important
     // when using controlled components within layers:
     // https://github.com/facebook/react/issues/1698
     // Then we restore state of any controlled component.
-    isNestingBatched = false;
-    restoreStateIfNeeded();
+    isBatching = false;
+    const controlledComponentsHavePendingUpdates = needsStateRestore();
+    if (controlledComponentsHavePendingUpdates) {
+      // We need to flush updates scheduled by controlled components before
+      // yielding back to the browser, so that the DOM is consistent before
+      // another event has the chance to fire.
```

Comment:
This comment doesn't make sense. There is no other event that will fire, and it is weird that `needsStateRestore` determines this.

We need to flush this before restoring state since we need to know if we should restore it or not.

Which also happens to accidentally give us the ability to flush changes before the browser event.

Reply:
You're right, I guess my comment is more about interactive updates in general, not controlled updates specifically. I'll change it to discuss `restoreState` instead.